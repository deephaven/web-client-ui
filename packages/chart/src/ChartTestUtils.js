import dh from '@deephaven/jsapi-shim';

class ChartTestUtils {
  static DEFAULT_CHART_TITLE = 'Chart Title';

  static DEFAULT_X_TITLE = 'X Axis';

  static DEFAULT_Y_TITLE = 'Y Axis';

  static DEFAULT_SERIES_NAME = 'MySeries';

  static makeAxis({
    label = 'Axis',
    type = dh.plot.AxisType.X,
    position = dh.plot.AxisPosition.BOTTOM,
    formatType = dh.Axis.FORMAT_TYPE_NUMBER,
    formatPattern = '###,###0.00',
    log = false,
  } = {}) {
    return new dh.Axis({
      label,
      type,
      position,
      formatType,
      formatPattern,
      log,
    });
  }

  static makeDefaultAxes() {
    return [
      ChartTestUtils.makeAxis({
        label: ChartTestUtils.DEFAULT_X_TITLE,
        type: dh.plot.AxisType.X,
      }),
      ChartTestUtils.makeAxis({
        label: ChartTestUtils.DEFAULT_Y_TITLE,
        type: dh.plot.AxisType.Y,
      }),
    ];
  }

  static makeSource({ axis = ChartTestUtils.makeAxis() }) {
    return new dh.SeriesDataSource({ axis, type: axis.type });
  }

  static makeDefaultSources() {
    const axes = ChartTestUtils.makeDefaultAxes();
    return axes.map(axis => ChartTestUtils.makeSource({ axis }));
  }

  static makeSeries({
    name = ChartTestUtils.DEFAULT_SERIES_NAME,
    plotStyle = dh.plot.SeriesPlotStyle.SCATTER,
    sources = ChartTestUtils.makeDefaultSources(),
    lineColor = null,
    shapeColor = null,
  } = {}) {
    return new dh.Series(name, plotStyle, sources, lineColor, shapeColor);
  }

  static makeChart({
    title = ChartTestUtils.DEFAULT_CHART_TITLE,
    series = [ChartTestUtils.makeSeries()],
    axes = ChartTestUtils.makeDefaultAxes(),
  } = {}) {
    return new dh.Chart({ title, series, axes });
  }

  static makeFigure({
    title = 'Figure',
    charts = [ChartTestUtils.makeChart()],
  } = {}) {
    return new dh.plot.Figure({ title, charts });
  }
}

export default ChartTestUtils;
