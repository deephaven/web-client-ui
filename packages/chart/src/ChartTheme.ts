import {
  getExpressionRanges,
  resolveCssVariablesInRecord,
} from '@deephaven/components';
import Log from '@deephaven/log';
import { ColorUtils } from '@deephaven/utils';
import chartThemeRaw from './ChartTheme.module.scss';

const log = Log.module('ChartTheme');

export interface ChartTheme {
  paper_bgcolor: string;
  plot_bgcolor: string;
  title_color: string;
  legend_color: string;
  colorway: string;
  gridcolor: string;
  linecolor: string;
  zerolinecolor: string;
  activecolor: string;
  rangebgcolor: string;
  area_color: string;
  trend_color: string;
  line_color: string;
  error_band_line_color: string;
  error_band_fill_color: string;
  ohlc_increasing: string;
  ohlc_decreasing: string;
}

export function defaultChartTheme(): Readonly<ChartTheme> {
  const chartTheme = resolveCssVariablesInRecord(chartThemeRaw);

  // The color normalization in `resolveCssVariablesInRecord` won't work for
  // colorway since it is an array of colors. We need to explicitly normalize
  // each color expression
  chartTheme.colorway = getExpressionRanges(chartTheme.colorway ?? '')
    .map(([start, end]) =>
      ColorUtils.normalizeCssColor(
        chartTheme.colorway.substring(start, end + 1)
      )
    )
    .join(' ');

  log.debug2('Chart theme:', chartThemeRaw);
  log.debug2('Chart theme derived:', chartTheme);

  return Object.freeze({
    paper_bgcolor: chartTheme['paper-bgcolor'],
    plot_bgcolor: chartTheme['plot-bgcolor'],
    title_color: chartTheme['title-color'],
    legend_color: chartTheme['legend-color'],
    colorway: chartTheme.colorway,
    gridcolor: chartTheme.gridcolor,
    linecolor: chartTheme.linecolor,
    zerolinecolor: chartTheme.zerolinecolor,
    activecolor: chartTheme.activecolor,
    rangebgcolor: chartTheme.rangebgcolor,
    area_color: chartTheme['area-color'],
    trend_color: chartTheme['trend-color'],
    line_color: chartTheme['line-color'],
    error_band_line_color: chartTheme['error-band-line-color'],
    error_band_fill_color: chartTheme['error-band-fill-color'],
    ohlc_increasing: chartTheme['ohlc-increasing'],
    ohlc_decreasing: chartTheme['ohlc-decreasing'],
  });
}

export default defaultChartTheme;
