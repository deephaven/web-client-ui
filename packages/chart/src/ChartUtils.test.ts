import dh from '@deephaven/jsapi-shim';
import { Formatter } from '@deephaven/jsapi-utils';
import { Layout } from 'plotly.js';
import ChartUtils from './ChartUtils';
import ChartTestUtils from './ChartTestUtils';
import ChartTheme from './ChartTheme';

const chartUtils = new ChartUtils(dh);

function makeFormatter() {
  return new Formatter();
}

it('groups the axes by type properly', () => {
  const testAxes = (axes, expectedResult) => {
    const chart = { axes };
    const result = ChartUtils.groupArray(chart.axes, 'type');
    expect(result).toEqual(expectedResult);
  };

  const xAxis1 = { type: 'x', label: 'x1' };
  // const xAxis2 = { type: 'x', label: 'x2' };
  // const xAxis3 = { type: 'x', label: 'x3' };
  const yAxis1 = { type: 'y', label: 'y1' };
  // const yAxis2 = { type: 'y', label: 'y3' };
  // const yAxis3 = { type: 'y', label: 'y3' };

  testAxes([], new Map());
  testAxes(
    [xAxis1, yAxis1],
    new Map([
      ['x', [xAxis1]],
      ['y', [yAxis1]],
    ])
  );
});

it('returns a newly composed layout object each time', () => {
  const layout1 = chartUtils.makeDefaultLayout(ChartTheme);
  const layout2 = chartUtils.makeDefaultLayout(ChartTheme);

  expect(layout1).not.toBe(layout2);
  expect(layout1.xaxis).not.toBe(layout2.xaxis);
  expect(layout1.yaxis).not.toBe(layout2.yaxis);
});

describe('date format string translations', () => {
  function testFormatString(
    format,
    expected,
    columnType = 'io.deephaven.db.tables.utils.DBDateTime',
    formatter = makeFormatter()
  ) {
    expect(
      chartUtils.getPlotlyDateFormat(formatter, columnType, format)
    ).toEqual(expect.objectContaining({ tickformat: expected }));
  }

  it('converts date format strings properly', () => {
    testFormatString('yyyy-MM-dd', '%Y-%m-%d');
    testFormatString('yyyy-MMM-dd', '%Y-%b-%d');
    testFormatString('yyyy-MMMM-dd', '%Y-%B-%d');
    testFormatString('yy-MM-dd', '%y-%m-%d');
    testFormatString('MM-dd-yyyy', '%m-%d-%Y');
    testFormatString('HH:mm:ss', '%H:%M:%S');
    testFormatString('HH:mm:ss.SSS', '%H:%M:%S.%3f');
    testFormatString('HH:mm:ss.SSSSSSSSS', '%H:%M:%S.%9f');
    testFormatString('yyyy-MM-dd hh:mm:ss', '%Y-%m-%d %I:%M:%S');
    testFormatString('yyyy-MM-dd HH:mm:ss', '%Y-%m-%d %H:%M:%S');
    testFormatString('yyyy-MM-dd HH:mm:ss.SSS', '%Y-%m-%d %H:%M:%S.%3f');
    testFormatString('yyyy-MM-dd HH:mm:ss.SSSSSSSSS', '%Y-%m-%d %H:%M:%S.%9f');
    testFormatString('M/d h:mm', '%-m/%-d %-I:%M');
  });

  it('converts day of week properly', () => {
    testFormatString('E', '%a');
    testFormatString('EE', '%a');
    testFormatString('EEE', '%a');
    testFormatString('EEEE', '%A');
    testFormatString('EEEEE', '%A');
    testFormatString('EEEEEE', '%A');
  });

  it('strips out the timezone properly', () => {
    testFormatString('yyyy-MM-dd z', '%Y-%m-%d');
    testFormatString('yy-MM-dd z', '%y-%m-%d');
  });

  it('converts the T separator properly', () => {
    testFormatString("yyyy-MM-dd'T'HH:mm:ss", '%Y-%m-%dT%H:%M:%S');
  });
});

describe('number format string translations', () => {
  function testFormatStringDeep(
    format,
    expected,
    columnType = 'java.lang.Float',
    formatter = makeFormatter()
  ) {
    expect(
      ChartUtils.getPlotlyNumberFormat(formatter, columnType, format)
    ).toEqual(expect.objectContaining(expected));
  }

  function testFormatString(
    format,
    expected,
    columnType = 'java.lang.Float',
    formatter = makeFormatter()
  ) {
    testFormatStringDeep(
      format,
      { tickformat: expected },
      columnType,
      formatter
    );
  }

  it('converts number format strings properly', () => {
    testFormatString('###,##0', '01,.0f');
    testFormatString('###,##0.00', '01,.2f');
    testFormatString('###,##0.0000', '01,.4f');

    // IDS-4565 there's an issue in d3 with trimming insignificant digits and specifying floating points:
    // https://github.com/d3/d3-format/issues/80
    // Re-enable this test case with the '~' trim option when fixed
    // testFormatString('#.####', '00.4~f');
    testFormatString('#.####', '00.4f');
  });

  it('handles % properly', () => {
    testFormatString('##0.00%', '01.2%');
  });

  it('handles exponential notation properly', () => {
    testFormatString('0.00E00', '01.2e');
  });

  it('handles the currency prefix properly', () => {
    testFormatStringDeep('$###,##0.00', {
      tickformat: '01,.2f',
      tickprefix: '$',
    });
  });

  it('handles the currency symbol properly', () => {
    testFormatStringDeep('\u00A4###,##0.00', {
      tickformat: '01,.2f',
      tickprefix: '$',
    });
    testFormatStringDeep('\u00A4\u00A4###,##0.00', {
      tickformat: '01,.2f',
      tickprefix: 'USD',
    });
    testFormatStringDeep('###,##0.00\u00A4', {
      tickformat: '01,.2f',
      ticksuffix: '$',
    });
    testFormatStringDeep('###,##0.00\u00A4\u00A4', {
      tickformat: '01,.2f',
      ticksuffix: 'USD',
    });
  });

  it('handles ignores negative subpatterns', () => {
    // IDS-4565 No way to directly translate negative subpattern with currently version of plot.ly
    // Just parse the positive subpattern for now
    testFormatStringDeep('00.00;-(00.00)', {
      tickformat: '02.2f',
      ticksuffix: '',
    });
  });
});

describe('updating layout axes', () => {
  function makeTwinAxes() {
    return [
      ChartTestUtils.makeAxis({ label: 'X Axis' }),
      ChartTestUtils.makeAxis({
        label: 'Y Axis',
        type: dh.plot.AxisType.Y,
        position: dh.plot.AxisPosition.RIGHT,
      }),
      ChartTestUtils.makeAxis({
        label: 'Y2 Axis',
        type: dh.plot.AxisType.Y,
        position: dh.plot.AxisPosition.LEFT,
      }),
    ];
  }

  it('adds new axes', () => {
    const layout = {};
    const axes = makeTwinAxes();
    chartUtils.updateLayoutAxes(layout, axes, axes);
    expect(layout).toEqual(
      expect.objectContaining({
        xaxis: expect.objectContaining({
          title: expect.objectContaining({ text: axes[0].label }),
          side: 'bottom',
        }),
        yaxis: expect.objectContaining({
          title: expect.objectContaining({ text: axes[1].label }),
          side: 'right',
        }),
        yaxis2: expect.objectContaining({
          title: expect.objectContaining({ text: axes[2].label }),
          side: 'left',
        }),
      })
    );
  });

  it('keeps the same axis objects, updates domain', () => {
    const layout: Partial<Layout> = {};
    const axes = makeTwinAxes();
    chartUtils.updateLayoutAxes(layout, axes, axes, 10);

    const { xaxis } = layout;
    const xDomain = [...(xaxis?.domain ?? [])];
    chartUtils.updateLayoutAxes(layout, axes, axes, 1000);

    expect(layout.xaxis).toBe(xaxis);
    expect(xDomain).not.toBe(xaxis?.domain);
  });

  it('removes stale axes', () => {
    const layout = {};
    const axes = makeTwinAxes();
    const chart = ChartTestUtils.makeChart({ axes });
    const figure = ChartTestUtils.makeFigure({ charts: [chart] });
    chartUtils.updateFigureAxes(layout, figure);
    expect(layout).toEqual(
      expect.objectContaining({
        xaxis: expect.objectContaining({}),
        yaxis: expect.objectContaining({}),
        yaxis2: expect.objectContaining({}),
      })
    );

    axes.pop();
    chartUtils.updateFigureAxes(layout, figure);
    expect(layout).toEqual(
      expect.objectContaining({
        xaxis: expect.objectContaining({}),
        yaxis: expect.objectContaining({}),
      })
    );
    expect(layout).not.toEqual(
      expect.objectContaining({
        yaxis2: expect.objectContaining({}),
      })
    );
  });

  describe('multiple axes', () => {
    const axes = [
      ChartTestUtils.makeAxis({ label: 'X Axis' }),
      ChartTestUtils.makeAxis({
        label: 'Y Axis',
        type: dh.plot.AxisType.Y,
        position: dh.plot.AxisPosition.RIGHT,
      }),
      ChartTestUtils.makeAxis({
        label: 'Y2 Axis',
        type: dh.plot.AxisType.Y,
        position: dh.plot.AxisPosition.RIGHT,
      }),
      ChartTestUtils.makeAxis({
        label: 'Y3 Axis',
        type: dh.plot.AxisType.Y,
        position: dh.plot.AxisPosition.RIGHT,
      }),
    ];
    const chart = ChartTestUtils.makeChart({ axes });
    const figure = ChartTestUtils.makeFigure({ charts: [chart] });
    const width = 400;
    const height = 1000;
    const expectedDomain = [0, 0.5];

    it('gets the chart bounds correctly', () => {
      const bounds = chartUtils.getChartBounds(figure, chart, width, height);
      expect(bounds).toEqual(
        expect.objectContaining({
          bottom: 0,
          left: 0,
          right: 0.875,
          top: 1,
        })
      );
    });

    it('positions multiple axes on the same side correctly', () => {
      const layout = {};
      const bounds = chartUtils.getChartBounds(figure, chart, width, height);
      const figureAxes = ChartUtils.getAllAxes(figure);
      chartUtils.updateLayoutAxes(
        layout,
        axes,
        figureAxes,
        width,
        height,
        bounds
      );
      expect(layout).toEqual(
        expect.objectContaining({
          xaxis: expect.objectContaining({
            domain: expect.arrayContaining(expectedDomain),
          }),
          yaxis: expect.objectContaining({}),
          yaxis2: expect.objectContaining({ anchor: 'free', position: 0.6875 }),
          yaxis3: expect.objectContaining({ anchor: 'free', position: 0.875 }),
        })
      );
    });
  });
});

describe('handles subplots and columns/rows correctly', () => {
  const width = ChartUtils.AXIS_SIZE_PX * 5;
  const height = ChartUtils.AXIS_SIZE_PX * 10;
  const halfXMargin = ChartUtils.AXIS_SIZE_PX / width / 2;
  const halfYMargin = ChartUtils.AXIS_SIZE_PX / height / 2;

  it('handles row location correctly', () => {
    const axes = ChartTestUtils.makeDefaultAxes();
    const charts = [
      ChartTestUtils.makeChart({ axes, row: 0 }),
      ChartTestUtils.makeChart({ axes, row: 1 }),
    ];
    const figure = ChartTestUtils.makeFigure({ charts, rows: 2 });
    expect(
      chartUtils.getChartBounds(figure, charts[0], width, height)
    ).toEqual({ bottom: 0.5 + halfYMargin, top: 1, left: 0, right: 1 });
    expect(
      chartUtils.getChartBounds(figure, charts[1], width, height)
    ).toEqual({ bottom: 0, top: 0.5 - halfYMargin, left: 0, right: 1 });
  });

  it('handles column location correctly', () => {
    const axes = ChartTestUtils.makeDefaultAxes();
    const charts = [
      ChartTestUtils.makeChart({ axes, column: 0 }),
      ChartTestUtils.makeChart({ axes, column: 1 }),
    ];
    const figure = ChartTestUtils.makeFigure({ charts, cols: 2 });
    expect(
      chartUtils.getChartBounds(figure, charts[0], width, height)
    ).toEqual({ bottom: 0, top: 1, left: 0, right: 0.5 - halfXMargin });
    expect(
      chartUtils.getChartBounds(figure, charts[1], width, height)
    ).toEqual({ bottom: 0, top: 1, left: 0.5 + halfXMargin, right: 1 });
  });

  it('handles colspan', () => {
    const axes = ChartTestUtils.makeDefaultAxes();
    const charts = [
      ChartTestUtils.makeChart({ axes, column: 0 }),
      ChartTestUtils.makeChart({ axes, column: 1 }),
      ChartTestUtils.makeChart({ axes, row: 1, colspan: 2 }),
    ];
    const figure = ChartTestUtils.makeFigure({ charts, cols: 2, rows: 2 });
    expect(chartUtils.getChartBounds(figure, charts[0], width, height)).toEqual(
      {
        bottom: 0.5 + halfYMargin,
        top: 1,
        left: 0,
        right: 0.5 - halfXMargin,
      }
    );
    expect(chartUtils.getChartBounds(figure, charts[1], width, height)).toEqual(
      {
        bottom: 0.5 + halfYMargin,
        top: 1,
        left: 0.5 + halfXMargin,
        right: 1,
      }
    );
    expect(
      chartUtils.getChartBounds(figure, charts[2], width, height)
    ).toEqual({ bottom: 0, top: 0.5 - halfYMargin, left: 0, right: 1 });
  });

  it('handles rowspan', () => {
    const axes = ChartTestUtils.makeDefaultAxes();
    const charts = [
      ChartTestUtils.makeChart({ axes, row: 0 }),
      ChartTestUtils.makeChart({ axes, row: 1 }),
      ChartTestUtils.makeChart({ axes, column: 1, rowspan: 2 }),
    ];
    const figure = ChartTestUtils.makeFigure({ charts, cols: 2, rows: 2 });
    expect(chartUtils.getChartBounds(figure, charts[0], width, height)).toEqual(
      {
        bottom: 0.5 + halfYMargin,
        top: 1,
        left: 0,
        right: 0.5 - halfXMargin,
      }
    );
    expect(chartUtils.getChartBounds(figure, charts[1], width, height)).toEqual(
      {
        bottom: 0,
        top: 0.5 - halfYMargin,
        left: 0,
        right: 0.5 - halfXMargin,
      }
    );
    expect(
      chartUtils.getChartBounds(figure, charts[2], width, height)
    ).toEqual({ bottom: 0, top: 1, left: 0.5 + halfXMargin, right: 1 });
  });
});

describe('returns the axis layout ranges properly', () => {
  function makeLayout(layout) {
    return {
      ...chartUtils.makeDefaultLayout(ChartTheme),
      ...layout,
    };
  }
  function testRange(layout, ranges) {
    expect(ChartUtils.getLayoutRanges(makeLayout(layout))).toEqual(ranges);
  }

  const xaxis = chartUtils.makeLayoutAxis(dh.plot.AxisType.X);
  xaxis.range = [0, 1];
  const xaxis2 = chartUtils.makeLayoutAxis(dh.plot.AxisType.X);
  xaxis2.range = [2, 3];
  const yaxis = chartUtils.makeLayoutAxis(dh.plot.AxisType.Y);
  yaxis.range = [4, 5];
  const yaxis2 = chartUtils.makeLayoutAxis(dh.plot.AxisType.Y);
  yaxis2.range = [6, 7];

  it('handles empty', () => {
    testRange({}, {});
  });
  it('handles one x-axis', () => {
    testRange({ xaxis }, { xaxis: xaxis.range });
  });
  it('handles x/y-axis', () => {
    testRange({ xaxis, yaxis }, { xaxis: xaxis.range, yaxis: yaxis.range });
  });
  it('handles multiple y-axes', () => {
    testRange(
      { xaxis, yaxis, yaxis2 },
      { xaxis: xaxis.range, yaxis: yaxis.range, yaxis2: yaxis2.range }
    );
  });
  it('handles multiple x/y-axes', () => {
    testRange(
      { xaxis, xaxis2, yaxis, yaxis2 },
      {
        xaxis: xaxis.range,
        xaxis2: xaxis2.range,
        yaxis: yaxis.range,
        yaxis2: yaxis2.range,
      }
    );
  });
});

it('converts a period time to decimal correctly', () => {
  expect(ChartUtils.periodToDecimal('06:00')).toBe(6);
  expect(ChartUtils.periodToDecimal('09:30:')).toBe(9.5);
  expect(ChartUtils.periodToDecimal('12:15')).toBe(12.25);
  expect(ChartUtils.periodToDecimal('15:45')).toBe(15.75);
  expect(ChartUtils.periodToDecimal('21:00')).toBe(21);
});

it('creates correct bounds from business days', () => {
  expect(
    chartUtils.createBoundsFromDays([
      'MONDAY',
      'TUESDAY',
      'WEDNESDAY',
      'THURSDAY',
      'FRIDAY',
    ])
  ).toStrictEqual([[6, 1]]);
  expect(
    chartUtils.createBoundsFromDays([
      'TUESDAY',
      'WEDNESDAY',
      'THURSDAY',
      'FRIDAY',
    ])
  ).toStrictEqual([[6, 2]]);
  expect(
    chartUtils.createBoundsFromDays([
      'MONDAY',
      'TUESDAY',
      'WEDNESDAY',
      'THURSDAY',
    ])
  ).toStrictEqual([[5, 1]]);
  expect(
    chartUtils.createBoundsFromDays(['MONDAY', 'TUESDAY', 'THURSDAY', 'FRIDAY'])
  ).toStrictEqual([
    [6, 1],
    [3, 4],
  ]);
  expect(
    chartUtils.createBoundsFromDays([
      'MONDAY',
      'TUESDAY',
      'WEDNESDAY',
      'THURSDAY',
    ])
  ).toStrictEqual([[5, 1]]);
  expect(
    chartUtils.createBoundsFromDays(['MONDAY', 'WEDNESDAY', 'FRIDAY'])
  ).toStrictEqual([
    [6, 1],
    [2, 3],
    [4, 5],
  ]);
  expect(
    chartUtils.createBoundsFromDays(['WEDNESDAY', 'THURSDAY'])
  ).toStrictEqual([[5, 3]]);
});

it('creates range breaks from holidays correctly', () => {
  const holidays = [
    new dh.Holiday('2020-06-22', []),
    new dh.Holiday('2020-06-22', [new dh.BusinessPeriod('10:00', '14:00')]),
    new dh.Holiday('2020-08-23', []),
    new dh.Holiday('2020-03-12', [
      new dh.BusinessPeriod('07:00', '08:00'),
      new dh.BusinessPeriod('21:00', '22:00'),
    ]),
  ];
  expect(chartUtils.createRangeBreakValuesFromHolidays(holidays)).toStrictEqual(
    [
      { values: ['2020-06-22 00:00:00.000000', '2020-08-23 00:00:00.000000'] },
      {
        dvalue: 36000000,
        values: ['2020-06-22 00:00:00.000000'],
      },
      {
        dvalue: 36000000,
        values: ['2020-06-22 14:00:00.000000'],
      },
      {
        dvalue: 25200000,
        values: ['2020-03-12 00:00:00.000000'],
      },
      {
        dvalue: 46800000,
        values: ['2020-03-12 08:00:00.000000'],
      },
      {
        dvalue: 7200000,
        values: ['2020-03-12 22:00:00.000000'],
      },
    ]
  );
});

describe('axis property name', () => {
  it('gets x/y axis property names correctly', () => {
    expect(chartUtils.getAxisPropertyName(dh.plot.AxisType.X)).toBe('x');
    expect(chartUtils.getAxisPropertyName(dh.plot.AxisType.Y)).toBe('y');
  });

  it('returns null for all other properties', () => {
    expect(chartUtils.getAxisPropertyName(dh.plot.AxisType.COLOR)).toBe(null);
    expect(chartUtils.getAxisPropertyName(dh.plot.AxisType.LABEL)).toBe(null);
    expect(chartUtils.getAxisPropertyName(dh.plot.AxisType.SHAPE)).toBe(null);
    expect(chartUtils.getAxisPropertyName(dh.plot.AxisType.SIZE)).toBe(null);
    expect(chartUtils.getAxisPropertyName(dh.plot.AxisType.Z)).toBe(null);
  });
});

describe('getPlotlyChartMode', () => {
  const { LINE, SCATTER } = dh.plot.SeriesPlotStyle;
  test.each([
    [null, null, 'markers'],
    [null, false, undefined],
    [null, true, 'markers'],
    [false, null, 'markers'],
    [false, false, undefined],
    [false, true, 'markers'],
    [true, null, 'lines+markers'],
    [true, false, 'lines'],
    [true, true, 'lines+markers'],
  ])(
    'returns the expected scatter plotly type for %s %s %s',
    (areLinesVisible, areShapesVisible, result) => {
      expect(
        chartUtils.getPlotlyChartMode(
          SCATTER,
          areLinesVisible,
          areShapesVisible
        )
      ).toBe(result);
    }
  );

  test.each([
    [null, null, 'lines'],
    [null, false, 'lines'],
    [null, true, 'lines+markers'],
    [false, null, undefined],
    [false, false, undefined],
    [false, true, 'markers'],
    [true, null, 'lines'],
    [true, false, 'lines'],
    [true, true, 'lines+markers'],
  ])(
    'returns the expected line plotly type for %s %s %s',
    (areLinesVisible, areShapesVisible, result) => {
      expect(
        chartUtils.getPlotlyChartMode(LINE, areLinesVisible, areShapesVisible)
      ).toBe(result);
    }
  );
});

describe('getMarkerSymbol', () => {
  const { getMarkerSymbol } = ChartUtils;
  it('returns valid shapes', () => {
    expect(getMarkerSymbol('SQUARE')).toBe('square');
    expect(getMarkerSymbol('CIRCLE')).toBe('circle');
    expect(getMarkerSymbol('DIAMOND')).toBe('diamond');
    expect(getMarkerSymbol('UP_TRIANGLE')).toBe('triangle-up');
    expect(getMarkerSymbol('DOWN_TRIANGLE')).toBe('triangle-down');
    expect(getMarkerSymbol('RIGHT_TRIANGLE')).toBe('triangle-right');
    expect(getMarkerSymbol('DOWN_TRIANGLE')).toBe('triangle-down');
  });

  it('throws on invalid shapes', () => {
    expect(() => getMarkerSymbol('ELLIPSE')).toThrow();
    expect(() => getMarkerSymbol('HORIZONTAL_RECTANGLE')).toThrow();
    expect(() => getMarkerSymbol('VERTICAL_RECTANGLE')).toThrow();
    expect(() => getMarkerSymbol('GARBAGE')).toThrow();
    expect(() => getMarkerSymbol('')).toThrow();
    expect(() => getMarkerSymbol('S')).toThrow();
    expect(() => getMarkerSymbol('&$*(#@&')).toThrow();
  });
});
