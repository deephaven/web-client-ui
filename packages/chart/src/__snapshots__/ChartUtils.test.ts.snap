// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeDefaultTemplate should create a default template 1`] = `
{
  "data": {
    "bar": [
      {
        "marker": {
          "line": {
            "color": "transparent",
          },
        },
      },
    ],
    "ohlc": [
      {
        "decreasing": {
          "line": {
            "color": "ChartTheme['ohlc_decreasing']",
          },
        },
        "increasing": {
          "line": {
            "color": "ChartTheme['ohlc_increasing']",
          },
        },
      },
    ],
    "pie": [
      {
        "outsidetextfont": {
          "color": "ChartTheme['title_color']",
        },
      },
    ],
    "scatter": [
      {
        "error_x": {
          "color": "ChartTheme['error_band_line_color']",
        },
        "error_y": {
          "color": "ChartTheme['error_band_line_color']",
        },
      },
    ],
    "treemap": [
      {
        "outsidetextfont": {
          "color": "ChartTheme['title_color']",
        },
      },
    ],
  },
  "layout": {
    "autosize": true,
    "colorway": [
      "ChartTheme[colorway0]",
      "ChartTheme[colorway1]",
      "ChartTheme[colorway2]",
    ],
    "datarevision": 0,
    "font": {
      "color": "ChartTheme['title_color']",
      "family": "'Fira Sans', sans-serif",
    },
    "geo": {
      "bgcolor": "ChartTheme['paper_bgcolor']",
      "coastlinecolor": "ChartTheme['coastline_color']",
      "lakecolor": "ChartTheme['lake_color']",
      "landcolor": "ChartTheme['land_color']",
      "oceancolor": "ChartTheme['ocean_color']",
      "rivercolor": "ChartTheme['river_color']",
      "showcoastlines": true,
      "showframe": false,
      "showlakes": true,
      "showland": true,
      "showocean": true,
      "showrivers": true,
    },
    "legend": {
      "font": {
        "color": "ChartTheme['title_color']",
      },
    },
    "margin": {
      "b": 60,
      "l": 60,
      "pad": 0,
      "r": 50,
      "t": 30,
    },
    "paper_bgcolor": "ChartTheme['paper_bgcolor']",
    "plot_bgcolor": "ChartTheme['plot_bgcolor']",
    "polar": {
      "angularaxis": {
        "automargin": true,
        "gridcolor": "ChartTheme['gridcolor']",
        "legend": {
          "font": {
            "color": "ChartTheme['legend_color']",
          },
        },
        "linecolor": "ChartTheme['linecolor']",
        "rangeslider": {
          "visible": false,
        },
        "showline": true,
        "tickcolor": "ChartTheme['paper_bgcolor']",
        "tickfont": {
          "color": "ChartTheme['zerolinecolor']",
        },
        "ticklen": 5,
        "ticks": "outside",
        "title": {
          "font": {
            "color": "ChartTheme['title_color']",
          },
        },
      },
      "bgcolor": "ChartTheme['plot_bgcolor']",
      "radialaxis": {
        "automargin": true,
        "gridcolor": "ChartTheme['gridcolor']",
        "legend": {
          "font": {
            "color": "ChartTheme['legend_color']",
          },
        },
        "linecolor": "ChartTheme['linecolor']",
        "rangeslider": {
          "visible": false,
        },
        "showline": true,
        "tickcolor": "ChartTheme['paper_bgcolor']",
        "tickfont": {
          "color": "ChartTheme['zerolinecolor']",
        },
        "ticklen": 5,
        "ticks": "outside",
        "title": {
          "font": {
            "color": "ChartTheme['title_color']",
          },
        },
      },
    },
    "scene": {
      "xaxis": {
        "automargin": true,
        "gridcolor": "ChartTheme['gridcolor']",
        "legend": {
          "font": {
            "color": "ChartTheme['legend_color']",
          },
        },
        "linecolor": "ChartTheme['linecolor']",
        "rangeslider": {
          "visible": false,
        },
        "showgrid": true,
        "showline": true,
        "tickcolor": "ChartTheme['paper_bgcolor']",
        "tickfont": {
          "color": "ChartTheme['zerolinecolor']",
        },
        "ticklen": 5,
        "ticks": "outside",
        "title": {
          "font": {
            "color": "ChartTheme['title_color']",
          },
        },
      },
      "yaxis": {
        "automargin": true,
        "gridcolor": "ChartTheme['gridcolor']",
        "legend": {
          "font": {
            "color": "ChartTheme['legend_color']",
          },
        },
        "linecolor": "ChartTheme['linecolor']",
        "rangeslider": {
          "visible": false,
        },
        "showline": true,
        "tickcolor": "ChartTheme['paper_bgcolor']",
        "tickfont": {
          "color": "ChartTheme['zerolinecolor']",
        },
        "ticklen": 5,
        "ticks": "outside",
        "title": {
          "font": {
            "color": "ChartTheme['title_color']",
          },
        },
        "zerolinecolor": "ChartTheme['zerolinecolor']",
        "zerolinewidth": 2,
      },
      "zaxis": {
        "automargin": true,
        "gridcolor": "ChartTheme['gridcolor']",
        "legend": {
          "font": {
            "color": "ChartTheme['legend_color']",
          },
        },
        "linecolor": "ChartTheme['linecolor']",
        "rangeslider": {
          "visible": false,
        },
        "showline": true,
        "tickcolor": "ChartTheme['paper_bgcolor']",
        "tickfont": {
          "color": "ChartTheme['zerolinecolor']",
        },
        "ticklen": 5,
        "ticks": "outside",
        "title": {
          "font": {
            "color": "ChartTheme['title_color']",
          },
        },
      },
    },
    "title": {
      "font": {
        "color": "ChartTheme['title_color']",
      },
      "pad": {
        "t": 8,
      },
      "xanchor": "center",
      "xref": "paper",
      "y": 1,
      "yanchor": "top",
    },
    "xaxis": {
      "automargin": true,
      "gridcolor": "ChartTheme['gridcolor']",
      "legend": {
        "font": {
          "color": "ChartTheme['legend_color']",
        },
      },
      "linecolor": "ChartTheme['linecolor']",
      "rangeslider": {
        "visible": false,
      },
      "showgrid": true,
      "showline": true,
      "tickcolor": "ChartTheme['paper_bgcolor']",
      "tickfont": {
        "color": "ChartTheme['zerolinecolor']",
      },
      "ticklen": 5,
      "ticks": "outside",
      "title": {
        "font": {
          "color": "ChartTheme['title_color']",
        },
      },
    },
    "yaxis": {
      "automargin": true,
      "gridcolor": "ChartTheme['gridcolor']",
      "legend": {
        "font": {
          "color": "ChartTheme['legend_color']",
        },
      },
      "linecolor": "ChartTheme['linecolor']",
      "rangeslider": {
        "visible": false,
      },
      "showline": true,
      "tickcolor": "ChartTheme['paper_bgcolor']",
      "tickfont": {
        "color": "ChartTheme['zerolinecolor']",
      },
      "ticklen": 5,
      "ticks": "outside",
      "title": {
        "font": {
          "color": "ChartTheme['title_color']",
        },
      },
      "zerolinecolor": "ChartTheme['zerolinecolor']",
      "zerolinewidth": 2,
    },
  },
}
`;
